dataset_preprocessing/image_matcher.py

```diff
@@ -0,0 +1,70 @@
+import numpy as np 
+import cv2
+from matplotlib import pyplot as plt
+import statistics
+import pandas as pd 
+from scipy import stats
+
+class ImageMatcher:
+    def __init__(self, baseImage):
+        self.baseImg = baseImage
+        self.searchImg = []
+        self.refineData, self.drawZone = False, False
+        self.allignmentStage = 0
+
+
+    def uploadSearchImage(self, searchImage):
+        if searchImage.any() == None:
+            print("image could not be found")
+            return False
+        self.searchImg = searchImage.copy()
+        return True 
+
+    def findObject(self, paddingX=0, paddingY=0, draw_box=False):
+        orb = cv2.ORB_create()
+        kp1, des1 = orb.detectAndCompute(self.baseImg,None)
+        kp2, des2 = orb.detectAndCompute(self.searchImg,None)
+    
+        bf = cv2.BFMatcher(cv2.NORM_HAMMING, crossCheck=True)
+        matches = bf.match(des1,des2)
+        matches = sorted(matches, key = lambda x:x.distance)
+        good_matches = matches[:10]
+        
+        pointsX = []
+        pointsY = []
+        for m in good_matches:
+            pointsX.append(kp2[m.trainIdx].pt[0])
+            pointsY.append(kp2[m.trainIdx].pt[1])
+
+        kp_img = cv2.drawMatches(self.baseImg,kp1,self.searchImg,kp2,good_matches,None,flags=cv2.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS)
+        self.searchImg, mask = self.__getBoundingBox(pointsX, pointsY, self.searchImg, paddingX, paddingY, draw_box)
+        
+        if mask is not None:
+            self.__fill_box(mask)
+
+        return kp_img, self.searchImg
+    
+    def __fill_box(self, mask):
```

Comment:
I think this should only be prefixed by one underscore.

Reply:
python private classes use two double underscores
https://www.geeksforgeeks.org/private-methods-in-python/#:~:text=Private%20methods%20are%20those%20methods,the%20double%20underscore%20%E2%80%9C__%E2%80%9D.

Reply:
[PEP 8](https://peps.python.org/pep-0008/#method-names-and-instance-variables) says:

> Use one leading underscore only for non-public methods and instance variables.
> ...
> Generally, double leading underscores should be used only to avoid name conflicts with attributes in classes designed to be subclassed.

Reply:
Wait that makes no sense though, python's built-in support for enforcing private vs public methods is literally double underlines. Not sure if maybe you can change it somehow through some setting, gonna keep the double underlines.

Reply:
It doesn't really matter to me lol, but python doesn't actually enforce private vs public stuff. It doesn't matter how many underscores we add technically we can always access it, it's more of a hint/convention thing I believe.

The double underscore is just to avoid name conflicts because:
> Python mangles these names with the class name: if class Foo has an attribute named __a, it cannot be accessed by Foo.__a. (An insistent user could still gain access by calling Foo._Foo__a.)